Gems/Atom/RHI/Code/Source/RHI/MultiDevicePipelineLibrary.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <Atom/RHI/Factory.h>
+#include <Atom/RHI/MultiDevicePipelineLibrary.h>
+#include <Atom/RHI/PipelineLibrary.h>
+#include <Atom/RHI/RHISystemInterface.h>
+#include <AzFramework/StringFunc/StringFunc.h>
+
+namespace AZ::RHI
+{
+    bool MultiDevicePipelineLibrary::ValidateIsInitialized() const
+    {
+        if (Validation::IsEnabled())
+        {
+            if (!IsInitialized())
+            {
+                AZ_Error(
+                    "MultiDevicePipelineLibrary",
+                    false,
+                    "MultiDevicePipelineLibrary is not initialized. This operation is only permitted on an initialized library.");
+                return false;
+            }
+        }
+        return true;
+    }
+
+    ResultCode MultiDevicePipelineLibrary::Init(MultiDevice::DeviceMask deviceMask, const MultiDevicePipelineLibraryDescriptor& descriptor)
+    {
+        if (Validation::IsEnabled())
+        {
+            if (IsInitialized())
+            {
+                AZ_Error(
+                    "MultiDevicePipelineLibrary",
+                    false,
+                    "MultiDevicePipelineLibrary is initialized. This operation is only permitted on an uninitialized library.");
+                return ResultCode::InvalidOperation;
+            }
+        }
+
+        MultiDeviceObject::Init(deviceMask);
+
+        ResultCode resultCode = ResultCode::Success;
+
+        IterateDevices(
+            [this, &descriptor, &resultCode](int deviceIndex)
+            {
+                auto* device = RHISystemInterface::Get()->GetDevice(deviceIndex);
+
+                m_devicePipelineLibraries[deviceIndex] = Factory::Get().CreatePipelineLibrary();
+                resultCode =
+                    m_devicePipelineLibraries[deviceIndex]->Init(*device, descriptor.GetDevicePipelineLibraryDescriptor(deviceIndex));
+
+                return resultCode == ResultCode::Success;
+            });
+
+        return resultCode;
```

Comment:
Do you need to clear m_devicePipelineLibraries incase some of devices were initialized successfully before the one which was failed?

Same may apply to all the other places which have similar logic

Reply:
Thanks for the suggestion, there would be multiple options to handle it, but I chose to go with your suggestion, i.e.
- As soon as one initialize fails, break from the loop
- Clear the map -> which shuts down already initialized resources
- Set the deviceMask to 0 such that the class is once again marked as uninitialized
- Report the error code to the caller